include/unifex/async_resource.hpp

```diff
@@ -0,0 +1,423 @@
+             });
+    });
+  }
+
+public:
+  _container_base(
+      void (*destruct_this_resource)(_container_base*) noexcept,
+      unifex::any_sender_of<> (*schedule_destruct)(
+          _container_base*, _container_base*) noexcept,
+      unifex::any_sender_of<> (*destroy_this_resource)(
+          _container_base*) noexcept,
+      void (*deleter)(_container_base*) noexcept) noexcept
+    : destruct_this_resource_(destruct_this_resource)
+    , schedule_destruct_(schedule_destruct)
+    , destroy_this_resource_(destroy_this_resource)
+    , deleter_(deleter)
+    , children_(nullptr)
+    , next_(nullptr)
+    , parent_(nullptr)
+    , state_(state_t::ALLOCATED) {}
+
+  state_t state() const noexcept {
+    // TODO optimize `memory_order_seq_cst`
+    return state_.load(std::memory_order_seq_cst);
+  }
+
+  unifex::task<void> close_scope() noexcept {
+    if (auto expected = state_t::CONSTRUCTED; state_.compare_exchange_strong(
+            expected, state_t::DESTROYING, std::memory_order_seq_cst)) {
+      destructionEvt_.set();
+      co_await unifex::when_all(scope_.get().complete(), close_child_scopes());
+    }
+  }
+
+  // happens when _ptr is dropped (non-nullptr)
+  // destroys from the root of tree post-order (objects created later might have
+  // references to objects created earlier, opposite is not true):
+  // 1. close all scopes
+  // 2. async destroy all resources
+  // 3. delete all containers
+  unifex::task<void> destroy() noexcept {
+    if (parent_) {
+      parent_->deregister_child(this);
+    }
+    co_await close_scope();
+    co_await destroy_resource();
+    destruct_container();
+  }
+
+  unifex::any_sender_of<> destruct(_container_base* parent) noexcept {
+    static_assert(
+        !unifex::sender_traits<decltype(destruct_impl())>::sends_done,
+        "destruct() must be unstoppable");
+    return schedule_destruct_(this, parent);
+  }
+
+  void register_child(_container_base* child) noexcept {
+    std::lock_guard lock{mutex_};
+    child->parent_ = this;
+    child->next_ = std::exchange(children_, child);
+  }
+
+  unifex::task<void> handle_construction_failure() noexcept {
+    switch (state()) {
+      case state_t::ALLOCATED: deleter_(this); break;
+      case state_t::DESTRUCT_SPAWNED: destructionEvt_.set(); break;
+      case state_t::CONSTRUCTED:
+        if (parent_ && parent_->state() == state_t::DESTROYING) {
+          co_await close_scope();
+          // bubble up cancellation, prevent returning an empty _ptr
+          co_await unifex::just_done();
+        }
+        break;
+      case state_t::DESTROYING:
+        // do nothing, ownership handed over to destroy()
+        break;
+    }
+  }
+};
+
+class async_scope_ref final {
+  unifex::v1::async_scope* scope_;
+  _container_base* container_;
+
+public:
+  /*implicit*/ async_scope_ref(unifex::v1::async_scope& scope) noexcept
+    : async_scope_ref(scope, nullptr) {}  // unmanaged scope
+
+  async_scope_ref(
```

Comment:
The style, inherited from the Committee's recommendations for the STL, is for this to be an `explicit` constructor.
```suggestion
  explicit async_scope_ref(
```